components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/OrientAsyncHelper.java

```diff
@@ -0,0 +1,143 @@
+ *
+ * @since 3.0
+ */
+public class OrientAsyncHelper
+    extends ComponentSupport
+{
+  private static final int DEFAULT_BUFFER_SIZE = 128;
+
+  private static final long DEFAULT_TIMEOUT_SECONDS = 60L;
+
+  private final int bufferSize;
+
+  private final long timeoutSeconds;
+
+  public OrientAsyncHelper() {
+    this(DEFAULT_BUFFER_SIZE, DEFAULT_TIMEOUT_SECONDS);
+  }
+
+  public OrientAsyncHelper(final int bufferSize, final long timeoutSeconds) {
+    checkArgument(bufferSize >= 1);
+    checkArgument(timeoutSeconds > 0);
+    this.bufferSize = bufferSize;
+    this.timeoutSeconds = timeoutSeconds;
+  }
+
+  /**
+   * Executes with passed in {@link ODatabaseDocumentTx} connection the passed in {@link String} SELECT SQL command
+   * and returns a true asynchronous backed {@link Iterable} of {@link ODocument} instance, that can be later
+   * transformed into whatever caller needs. This means that this wrapper supports ANY SQL commands, even those
+   * performing aggregation on results, as long as iterator is not accumulated into a real array or list.
+   */
+  public Iterable<ODocument> asyncIterable(final ODatabaseDocumentTx db,
+                                           final String selectQuery,
+                                           final Map<String, Object> parameters)
+  {
+    checkNotNull(db);
+    checkNotNull(selectQuery);
+    final ArrayBlockingQueue<ODocument> queue = new ArrayBlockingQueue<>(bufferSize);
+
+    db.command(new OSQLNonBlockingQuery<ODocument>(selectQuery, new OCommandResultListener()
+    {
+      @Override
+      public boolean result(final Object iRecord) {
+        final ODocument doc = (ODocument) iRecord;
+        try {
+          queue.offer(doc, timeoutSeconds, TimeUnit.SECONDS);
+        }
+        catch (InterruptedException e) {
+          throw Throwables.propagate(e);
+        }
+        return true;
+      }
+
+      @Override
+      public void end() {
+        try {
+          queue.offer(SENTINEL, timeoutSeconds, TimeUnit.SECONDS);
+        }
+        catch (InterruptedException e) {
+          throw Throwables.propagate(e);
+        }
+      }
+    })).execute(parameters);
+
+    return new OrientAsyncIterableWrapper(queue);
+  }
+
+  private static final ODocument SENTINEL = new ODocument();
+
+  private static final class OrientAsyncIterableWrapper
+      implements Iterable<ODocument>, Iterator<ODocument>
+  {
+    private final BlockingQueue<ODocument> queue;
+
+    private ODocument next;
+
+    public OrientAsyncIterableWrapper(final BlockingQueue<ODocument> queue)
+    {
+      this.queue = queue;
+    }
+
+    @Override
+    public Iterator<ODocument> iterator() {
+      return this;
+    }
+
+    @Override
+    public boolean hasNext() {
+      try {
+        next = queue.poll(DEFAULT_TIMEOUT_SECONDS, TimeUnit.SECONDS);
+      }
+      catch (InterruptedException e) {
+        throw Throwables.propagate(e);
+      }
+      return next != SENTINEL;
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Use equals() to compare object references. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=pmd%3ACompareObjectsWithEquals)